#!/usr/bin/env python

include: "master.smk"
include: "rules/00_transfer_ref_genome.smk"
include: "rules/01_softlink_fastq.smk"
include: "rules/01_download_sra.smk"
include: "rules/02_trim_fastq.smk"
include: "rules/03_align_fastq.smk"
include: "rules/03_bismark2report.smk"
include: "rules/04_deduplicate_bam.smk"
include: "rules/05_call_methylation.smk"
include: "rules/06_call_variants.smk"
include: "rules/07_calculate_statistics.smk"
include: "rules/08_methylseekr_and_TDF.smk"
include: "rules/09_majel_cleanup.smk"

localrules: all, softlink_fastq, calculate_conversion, cleanup, rsync

# Specify all output files resulting from Majel pipeline
# Dictionaries are derived from master.smk
local_log = os.path.join(config['output_path'], 'snakemake')
output_files = get_all_files(D_genomes, D_sample_details)

rule all:
    input:
        output_files

onstart:
    shell("cat {log} >> {local_log}.log")

onsuccess:
    copy_log_to_sample(log)
    shell("cat {log} >> {local_log}.log")

    for sample, sample_details in D_sample_details.items():
        remove_temp_files(sample, sample_details)

onerror:
    copy_log_to_sample(log)
    shell("cat {log} >> {local_log}.error")

'''
Rules:
    Snakefile:
        all                     - establishes all output files from all rules
    
    00_transfer_ref_genome.smk:
        transfer_ref_genome     - transfers reference genome files to working directory when using an HPC job scheduling system

    01_softlink_fastq.smk:
        softlink_fastq          - softlinks fastq files derived from "data_dir" and "file_prefixes" --config variable
    
    01_sra_download.smk:
        sra_download            - downloads run accessions stiulated in "file_prefixes --config variable
        sra_to_fastq            - converts sra files to fastqs
    
    02_trim_fastq.smk:
        move_umi                - moves umis from read to name uisng Fastp
        trim_fastq              - trims fastqs using trim_galore
        merge_fastq             - merges r1 fastqs together and r2 fastqs together
    
    03_align_fastq.smk:
        bismark_align           - aligns fastq sequences using Bismark to produce a BAM file
        sort_bam                - sorts BAM file produced by Bismark

    03_bismark2report.smk:
        bismark_deduplicate     - used only for bismark2report: deduplicates BAM file
        bismark_methylation     - used only for bismark2report: calls cytosine methylation
        bismark2report          - produces Bismark html report

    04_deduplicate_bam.smk:
        deduplicate_bam         - deduplicates reads in sorted BAM using Gencore for UMI support
        merge_deduplicate_bams  - merges deduplicated CT and GA BAMs

    05_call_methylation.smk:
        call_methylation        - calls cytosine methylation using MethylDackel for downstream analysis

    05_call_variants.smk:
        mask_converted_bases    - masks base positions in BAM potentially affected by cytosine conversion
        coverage_to_regions     - reates fa.fai file that splits genome into 500 regions based on coverage
        call_variants           - calls variants from cytosine converted data

    06_calculate_statistics.smk:
        calculate_coverage      - calculates sequencing coverage statistics
        calculate_conversion    - calculates cytosine conversion statistics

    07_methylseekr_and_TDF.smk:
        methylseekr_and_TDF     - calls UMRs and LMRs with and without PMDs, produces TDF file for IGV
'''
